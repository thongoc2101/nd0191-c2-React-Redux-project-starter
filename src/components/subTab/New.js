import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { QuestionActions } from "../../redux/actions/question.action";
import { setQuestionUser } from "../../redux/slice/auth.slice";
import { useNavigate } from "react-router-dom";

const New = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate();

    const { userInfo } = useSelector((state) => state.auth);
    const [firstOption, setFirstOption] = useState('');
    const [secondOption, setSecondOption] = useState('');

    const [showErrorMessage, setShowErrorMessage] = useState(false);

    const onSubmit = (e) => {
        e.preventDefault();

        // Check for unique question options. Both options should not be the same.
        if (firstOption === secondOption) {
            setShowErrorMessage(true);
            return;
        }

        dispatch(QuestionActions.addQuestion({
            optionOneText: firstOption,
            optionTwoText: secondOption,
            author: userInfo.id
        })).then((res) => {
            setFirstOption('');
            setSecondOption('');
            dispatch(setQuestionUser(res.payload));
            navigate({ pathname: '/' });
        });
    };

    const inputStyle = "w-full mt-2 px-3 py-2 border border-slate-300 rounded-md";
    return (
        <div className="new-page-container px-14">
            <h1 className="text-3xl font-bold mt-9">Would You Rather</h1>
            <h3 className="text-sm">Create Your Own Poll</h3>
            <form onSubmit={onSubmit}>

                <div className="mt-3">
                    <label htmlFor="firstOption"
                        className="block text-sm font-bold">First Option</label>
                    <input
                        value={firstOption}
                        onChange={(e) => setFirstOption(e.target.value)}
                        type="text"
                        name="firstOption"
                        id="firstOption"
                        maxLength={255}
                        className={inputStyle} />
                </div>

                <div className="mt-3">
                    <label htmlFor="secondOption"
                        className="block text-sm font-bold">Second Option</label>
                    <input
                        value={secondOption}
                        onChange={(e) => setSecondOption(e.target.value)}
                        type="text"
                        name="secondOption"
                        id="secondOption"
                        maxLength={255}
                        className={inputStyle} />
                </div>

                {showErrorMessage &&
                    <div className="mt-3 text-left">
                        <span className="text-red-600">Both options should not be the same</span>
                    </div>
                }

                <div className="mt-5 text-right">
                    <button type="submit" disabled={!firstOption || !secondOption}
                        className="bg-sky-500 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white disabled:bg-slate-500 disabled:text-white disabled:cursor-not-allowed">
                        Submit
                    </button>
                </div>

            </form>
        </div>
    );
}

export default New;