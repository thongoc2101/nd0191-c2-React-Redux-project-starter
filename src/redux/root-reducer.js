import { combineReducers } from "@reduxjs/toolkit";
import { AuthSlice } from "./slice/auth.slice";
import { QuestionSlice } from "./slice/question.slice";

export const rootReducer = combineReducers({
    auth: AuthSlice.reducer,
    questions: QuestionSlice.reducer,
});