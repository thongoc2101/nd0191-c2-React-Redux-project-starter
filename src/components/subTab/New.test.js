import {fireEvent, render} from "@testing-library/react";
import {Provider} from "react-redux";
import {store} from "../../redux/store";
import {BrowserRouter} from "react-router-dom";
import React from "react";
import New from "./New";

describe("New", () => {
    it("should render the component", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <New />
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it("should display all elements", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <New/>
                </BrowserRouter>
            </Provider>
        );

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();

        const firstOptionLabelElement = component.getByTestId("firstOptionLabel");
        const firstOptionInputElement = component.getByTestId("firstOption");
        const secondOptionLabelElement = component.getByTestId("secondOptionLabel");
        const secondOptionInputElement = component.getByTestId("secondOption");
        const submitButtonElement = component.getByTestId("submit-new-poll");

        expect(firstOptionLabelElement.textContent).toBe("First Option");
        expect(secondOptionLabelElement.textContent).toBe("Second Option");
        expect(submitButtonElement.textContent).toBe("Submit");

        fireEvent.change(firstOptionInputElement, {target: {value: 'Text 1'}});
        fireEvent.change(secondOptionInputElement, {target: {value: 'Text 2'}});
        expect(firstOptionInputElement.value).toBe("Text 1");
        expect(secondOptionInputElement.value).toBe("Text 2");
    });
});