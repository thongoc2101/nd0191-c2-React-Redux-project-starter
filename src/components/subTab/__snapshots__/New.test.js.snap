// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New should display all elements 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="new-page-container px-14"
      >
        <h1
          class="text-3xl font-bold mt-9"
        >
          Would You Rather
        </h1>
        <h3
          class="text-sm"
        >
          Create Your Own Poll
        </h3>
        <form>
          <div
            class="mt-3"
          >
            <label
              class="block text-sm font-bold"
              data-testid="firstOptionLabel"
              for="firstOption"
            >
              First Option
            </label>
            <input
              class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
              data-testid="firstOption"
              id="firstOption"
              maxlength="255"
              name="firstOption"
              type="text"
              value=""
            />
          </div>
          <div
            class="mt-3"
          >
            <label
              class="block text-sm font-bold"
              data-testid="secondOptionLabel"
              for="secondOption"
            >
              Second Option
            </label>
            <input
              class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
              data-testid="secondOption"
              id="secondOption"
              maxlength="255"
              name="secondOption"
              type="text"
              value=""
            />
          </div>
          <div
            class="mt-5 text-right"
          >
            <button
              class="bg-sky-500 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white disabled:bg-slate-500 disabled:text-white disabled:cursor-not-allowed"
              data-testid="submit-new-poll"
              disabled=""
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="new-page-container px-14"
    >
      <h1
        class="text-3xl font-bold mt-9"
      >
        Would You Rather
      </h1>
      <h3
        class="text-sm"
      >
        Create Your Own Poll
      </h3>
      <form>
        <div
          class="mt-3"
        >
          <label
            class="block text-sm font-bold"
            data-testid="firstOptionLabel"
            for="firstOption"
          >
            First Option
          </label>
          <input
            class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
            data-testid="firstOption"
            id="firstOption"
            maxlength="255"
            name="firstOption"
            type="text"
            value=""
          />
        </div>
        <div
          class="mt-3"
        >
          <label
            class="block text-sm font-bold"
            data-testid="secondOptionLabel"
            for="secondOption"
          >
            Second Option
          </label>
          <input
            class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
            data-testid="secondOption"
            id="secondOption"
            maxlength="255"
            name="secondOption"
            type="text"
            value=""
          />
        </div>
        <div
          class="mt-5 text-right"
        >
          <button
            class="bg-sky-500 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white disabled:bg-slate-500 disabled:text-white disabled:cursor-not-allowed"
            data-testid="submit-new-poll"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`New should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="new-page-container px-14"
      >
        <h1
          class="text-3xl font-bold mt-9"
        >
          Would You Rather
        </h1>
        <h3
          class="text-sm"
        >
          Create Your Own Poll
        </h3>
        <form>
          <div
            class="mt-3"
          >
            <label
              class="block text-sm font-bold"
              data-testid="firstOptionLabel"
              for="firstOption"
            >
              First Option
            </label>
            <input
              class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
              data-testid="firstOption"
              id="firstOption"
              maxlength="255"
              name="firstOption"
              type="text"
              value=""
            />
          </div>
          <div
            class="mt-3"
          >
            <label
              class="block text-sm font-bold"
              data-testid="secondOptionLabel"
              for="secondOption"
            >
              Second Option
            </label>
            <input
              class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
              data-testid="secondOption"
              id="secondOption"
              maxlength="255"
              name="secondOption"
              type="text"
              value=""
            />
          </div>
          <div
            class="mt-5 text-right"
          >
            <button
              class="bg-sky-500 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white disabled:bg-slate-500 disabled:text-white disabled:cursor-not-allowed"
              data-testid="submit-new-poll"
              disabled=""
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="new-page-container px-14"
    >
      <h1
        class="text-3xl font-bold mt-9"
      >
        Would You Rather
      </h1>
      <h3
        class="text-sm"
      >
        Create Your Own Poll
      </h3>
      <form>
        <div
          class="mt-3"
        >
          <label
            class="block text-sm font-bold"
            data-testid="firstOptionLabel"
            for="firstOption"
          >
            First Option
          </label>
          <input
            class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
            data-testid="firstOption"
            id="firstOption"
            maxlength="255"
            name="firstOption"
            type="text"
            value=""
          />
        </div>
        <div
          class="mt-3"
        >
          <label
            class="block text-sm font-bold"
            data-testid="secondOptionLabel"
            for="secondOption"
          >
            Second Option
          </label>
          <input
            class="w-full mt-2 px-3 py-2 border border-slate-300 rounded-md"
            data-testid="secondOption"
            id="secondOption"
            maxlength="255"
            name="secondOption"
            type="text"
            value=""
          />
        </div>
        <div
          class="mt-5 text-right"
        >
          <button
            class="bg-sky-500 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white disabled:bg-slate-500 disabled:text-white disabled:cursor-not-allowed"
            data-testid="submit-new-poll"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
